tool/tsh/common/vnet_common.go

```diff
@@ -18,35 +18,190 @@ package common
 func (p *vnetAppProvider) GetCachedClient(ctx context.Context, profileName, leafClusterName string) (*client.ClusterClient, error) {
+	return p.clientCache.Get(ctx, profileName, leafClusterName)
+}
+
+// ReissueAppCert returns a new app certificate for the given app in the named profile and leaf cluster.
+// It uses retryWithRelogin to issue the new app cert. A relogin may not be necessary if the app cert lifetime
+// was shorter than the cluster cert lifetime, or if the user has already re-logged in to the cluster.
+// If a cluster relogin is completed, the cluster client cache will be cleared for the root cluster and all
+// leaf clusters of that root.
+func (p *vnetAppProvider) ReissueAppCert(ctx context.Context, profileName, leafClusterName string, app types.Application) (tls.Certificate, error) {
+	tc, err := p.newTeleportClient(ctx, profileName, leafClusterName)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	var cert tls.Certificate
+	err = p.retryWithRelogin(ctx, tc, func() error {
+		var err error
+		cert, err = p.reissueAppCert(ctx, tc, profileName, leafClusterName, app)
+		return trace.Wrap(err, "reissuing app cert")
+	})
+	return cert, trace.Wrap(err)
+}
+
+// GetDialOptions returns ALPN dial options for the profile.
+func (p *vnetAppProvider) GetDialOptions(ctx context.Context, profileName string) (*vnet.DialOptions, error) {
+	profile, err := p.clientStore.GetProfile(profileName)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading user profile")
+	}
+	dialOpts := &vnet.DialOptions{
+		WebProxyAddr:            profile.WebProxyAddr,
+		ALPNConnUpgradeRequired: profile.TLSRoutingConnUpgradeRequired,
+	}
+	if dialOpts.ALPNConnUpgradeRequired {
+		dialOpts.RootClusterCACertPool, err = p.getRootClusterCACertPool(ctx, profileName)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return dialOpts, nil
+}
+
+// getRootClusterCACertPool returns a certificate pool for the root cluster of the given profile.
+func (p *vnetAppProvider) getRootClusterCACertPool(ctx context.Context, profileName string) (*x509.CertPool, error) {
+	tc, err := p.newTeleportClient(ctx, profileName, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "creating new client")
+	}
+	certPool, err := tc.RootClusterCACertPool(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading root cluster CA cert pool")
+	}
+	return certPool, nil
+}
+
+func (p *vnetAppProvider) retryWithRelogin(ctx context.Context, tc *client.TeleportClient, fn func() error, opts ...client.RetryWithReloginOption) error {
+	profileName, err := utils.Host(tc.WebProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	didLock := false
+	defer func() {
+		if didLock {
+			p.loginMu.Unlock()
+		}
+	}()
+
+	opts = append(opts,
+		client.WithBeforeLoginHook(func() error {
+			if p.loginMu.TryLock() {
+				didLock = true
+			} else {
+				return fmt.Errorf("Not attempting re-login to cluster %s, currently logging into another cluster.", tc.SiteName)
+			}
```

Comment:
I guess this warrants a comment, but if there was no try lock here, concurrent connections to apps from the same expired cluster would trigger `RetryWithLogin` multiple times. Once unlocked, each of them would perform relogin one after the other.

That's what I did a while ago in Connect. At some point I realized that a better UX would be to have an additional lock per profile, so that all those queued up attempts to relogin could just re-check after acquiring a lock if the cert got refreshed.

Reply:
I'm going a bit further than what you describe here and only allowing one login to any cluster at a time, and returning an error if a login is already in process. UX of logging in to multiple clusters (or the same cluster) at a time in tsh is pretty bad especially if mfa prompts are involved. But it looks like connect does the same thing https://github.com/gravitational/teleport/blob/92abf0101890bb89028713894a698d9401de91f1/lib/teleterm/daemon/daemon.go#L104-L109

Reply:
Just added some comments in the source